File: src/server/telemetry/sync_worker.go
package main

import (
	"bytes"
	"database/sql"
	"encoding/json"
	"log"
	"net/http"
	"os"
	"time"

	_ "github.com/mattn/go-sqlite3"
)

type Metric struct {
	ID         int       `json:"-"`
	MetricName string    `json:"metric_name"`
	MetricType string    `json:"metric_type"`
	Value      float32   `json:"value"`
	LabelsJSON string    `json:"-"`
	Labels     any       `json:"labels"`
	Timestamp  time.Time `json:"timestamp"`
}

func SyncMetricsToCloud(db *sql.DB, cloudURL string) error {
	rows, err := db.Query("SELECT id, metric_name, metric_type, value, labels_json, timestamp FROM local_telemetry_buffer WHERE sync_status = 'pending' LIMIT 100")
	if err != nil {
		return err
	}
	defer rows.Close()

	var metrics []Metric
	var ids []int
	for rows.Next() {
		var m Metric
		var ts string
		if err := rows.Scan(&m.ID, &m.MetricName, &m.MetricType, &m.Value, &m.LabelsJSON, &ts); err != nil {
			log.Printf("Failed to scan metric: %v", err)
			continue
		}

		parsedTime, err := time.Parse(time.RFC3339Nano, ts)
		if err == nil {
			m.Timestamp = parsedTime
		}

		var labels any
		json.Unmarshal([]byte(m.LabelsJSON), &labels)
		m.Labels = labels

		metrics = append(metrics, m)
		ids = append(ids, m.ID)
	}

	if len(metrics) == 0 {
		return nil
	}

	payload, err := json.Marshal(metrics)
	if err != nil {
		return err
	}

	resp, err := http.Post(cloudURL+"/api/telemetry/sync", "application/json", bytes.NewBuffer(payload))
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode == 200 {
		for _, id := range ids {
			db.Exec("UPDATE local_telemetry_buffer SET sync_status = 'synced' WHERE id = ?", id)
		}
		log.Printf("Successfully synced %d metrics to cloud", len(metrics))
	} else {
		log.Printf("Failed to sync metrics, status code: %d", resp.StatusCode)
	}

	return nil
}

func main() {
	dbURL := os.Getenv("DATABASE_URL")
	if dbURL == "" {
		dbURL = "sqlite://ohc-standalone.db"
	}

	// remove "sqlite://" prefix for go-sqlite3
	if len(dbURL) > 9 && dbURL[:9] == "sqlite://" {
		dbURL = dbURL[9:]
	}

	cloudURL := os.Getenv("CLOUD_URL")
	if cloudURL == "" {
		cloudURL = "https://cloud.onehumancorp.com"
	}

	db, err := sql.Open("sqlite3", dbURL)
	if err != nil {
		log.Fatalf("Failed to open database: %v", err)
	}
	defer db.Close()

	ticker := time.NewTicker(60 * time.Second)
	for range ticker.C {
		if err := SyncMetricsToCloud(db, cloudURL); err != nil {
			log.Printf("Error syncing telemetry: %v", err)
		}
	}
}

File: src/server/telemetry/sync_worker.rs
use sqlx::{query, SqlitePool, Row};
use chrono::{DateTime, Utc};
use serde_json::Value;

#[derive(Debug)]
pub struct Metric {
    pub id: i32,
    pub metric_name: String,
    pub metric_type: String,
    pub value: f32,
    pub labels_json: String,
    pub timestamp: DateTime<Utc>,
}

pub async fn sync_metrics_to_cloud(db: &SqlitePool, cloud_url: &str) -> Result<(), Box<dyn std::error::Error>> {
    let rows = query("SELECT id, metric_name, metric_type, value, labels_json, timestamp FROM local_telemetry_buffer WHERE sync_status = 'pending' LIMIT 100")
        .fetch_all(db)
        .await?;

    if rows.is_empty() {
        return Ok(());
    }

    let mut metrics = Vec::new();
    let mut ids = Vec::new();

    for row in rows {
        let id: i32 = row.get("id");
        let metric_name: String = row.get("metric_name");
        let metric_type: String = row.get("metric_type");
        let value: f32 = row.get("value");
        let labels_json: String = row.get("labels_json");

        let ts_str: String = row.get("timestamp");
        let timestamp = DateTime::parse_from_rfc3339(&ts_str)?.with_timezone(&Utc);

        let labels: Value = serde_json::from_str(&labels_json).unwrap_or(Value::Null);

        metrics.push(serde_json::json!({
            "metric_name": metric_name,
            "metric_type": metric_type,
            "value": value,
            "labels": labels,
            "timestamp": timestamp,
        }));
        ids.push(id);
    }

    let client = reqwest::Client::builder()
        .timeout(std::time::Duration::from_secs(10))
        .build()
        .unwrap_or_else(|_| reqwest::Client::new());

    let res = client.post(format!("{}/api/telemetry/sync", cloud_url))
        .json(&metrics)
        .send()
        .await;

    match res {
        Ok(response) => {
            if response.status().is_success() {
                for id in ids {
                    query("UPDATE local_telemetry_buffer SET sync_status = 'synced' WHERE id = $1")
                        .bind(id)
                        .execute(db)
                        .await?;
                }
            } else {
                tracing::error!("Failed to sync metrics, status code: {}", response.status());
            }
        },
        Err(e) => {
            tracing::error!("Error syncing telemetry: {}", e);
        }
    }

    Ok(())
}

pub fn start_telemetry_sync_worker(db: SqlitePool, cloud_url: String, interval: std::time::Duration) {
    tokio::spawn(async move {
        let mut ticker = tokio::time::interval(interval);
        loop {
            ticker.tick().await;
            if let Err(e) = sync_metrics_to_cloud(&db, &cloud_url).await {
                tracing::error!("Error in telemetry sync worker: {}", e);
            }
        }
    });
}
